File: Week2/김다빈/BOJ_18511_1/18511.py
# (10 ≤ N ≤ 100,000,000, 1 ≤ K의 원소의 개수 ≤ 3)

n, k = map(int, input().split())
nums = list(map(int, input().split()))
answer = ""


def backtracking(curNum, numArr, targetNum):
    global answer

    if len(curNum) > 0 and int(curNum) > int(targetNum):
        return

    if len(curNum) > 0 and int(targetNum) >= int(curNum):
        if answer == "" or int(curNum) > int(answer):
            answer = curNum

    for i in numArr:
        backtracking(curNum + str(i), numArr, targetNum)


backtracking(answer, nums, str(n))
print(answer)

File: Week2/김다빈/BOJ_18511_1/18511.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    private static String answer = "";
    private static ArrayList<Integer> nums = new ArrayList<>();
    private static String targetNum;

    private static void backtracking(String curNum) {
        if (!curNum.isEmpty() && Long.parseLong(curNum) > Long.parseLong(targetNum)) {
            return;
        }

        if (!curNum.isEmpty()) {
            if (answer.isEmpty() || Long.parseLong(curNum) > Long.parseLong(answer)) {
                answer = curNum;
            }
        }

        for (int num : nums) {
            backtracking(curNum + num);
        }

    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        targetNum = String.valueOf(n);
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < k; i++ ) {
            nums.add(Integer.parseInt(st.nextToken()));
        }

        backtracking("");
        System.out.println(answer);

    }
}